test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java

```diff
@@ -362,92 +364,79 @@ public final void testUnknownObjectException() throws IOException {
-        boolean withinQuotes = false;
-        boolean expectedException = true;
-        int objectHoldingArbitraryContentLevel = 0;
-        for (int insertionPosition = 0; insertionPosition < validQuery.length(); insertionPosition++) {
-            if (validQuery.charAt(insertionPosition) == '"') {
-                withinQuotes = withinQuotes == false;
-            } else if (withinQuotes == false && validQuery.charAt(insertionPosition) == '}') {
-                level--;
-                if (expectedException == false) {
-                    //track where we are within the object that holds arbitrary content
-                    objectHoldingArbitraryContentLevel--;
-                }
-                if (objectHoldingArbitraryContentLevel == 0) {
-                    //reset the flag once we have traversed the whole object that holds arbitrary content
-                    expectedException = true;
-                }
-            } else if (withinQuotes == false && validQuery.charAt(insertionPosition) == '{') {
-                //keep track of which level we are within the json so that we can properly close the additional object
-                level++;
-                //if we don't expect an exception, it means that we are within an object that can contain arbitrary content.
-                //in that case we ignore the whole object including its children, no need to even check where we are.
-                if (expectedException) {
-                    int startCurrentObjectName = -1;
-                    int endCurrentObjectName = -1;
-                    //look backwards for the current object name, to find out whether we expect an exception following its mutation
-                    for (int i = insertionPosition; i >= 0; i--) {
-                        if (validQuery.charAt(i) == '}') {
-                            break;
-                        } else if (validQuery.charAt(i) == '"') {
-                            if (endCurrentObjectName == -1) {
-                                endCurrentObjectName = i;
-                            } else if (startCurrentObjectName == -1) {
-                                startCurrentObjectName = i + 1;
-                            } else {
-                                break;
+        // Track the number of query mutations
+        int mutation = 0;
+
+        // Indicate if a part of the query can hold any arbitrary content
+        boolean hasArbitraryContent = getObjectsHoldingArbitraryContent().isEmpty() == false;
+
+        while(true) {
+            boolean expectedException = true;
+
+            BytesStreamOutput out = new BytesStreamOutput();
+            try (XContentGenerator generator = XContentType.JSON.xContent().createGenerator(out)) {
+                // Parse the valid query and inserts a new object level called "newField"
+                try (XContentParser parser = XContentHelper.createParser(new BytesArray(validQuery))) {
+                    int depth = -1;
```

Comment:
You got it :) `depth`was decreased every time `END_OBJECT` was found but I then removed this part. `objectIndex` sounds good.